code/blake2/Hacl.Impl.Blake2.Generic.fst

```diff
@@ -622,16 +622,23 @@ inline_for_extraction noextract
 val blake2_finish:#al:Spec.alg -> #ms:m_spec -> blake2_finish_t al ms
 
 let blake2_finish #al #ms nn output hash =
-  let h0 = ST.get () in
-  salloc1 h0 (2ul *. (size_row al)) (u8 0) (Ghost.hide (loc output))
-  (fun _ h1 -> live h1 output /\ h1.[|output|] == Spec.blake2_finish al (state_v h0 hash) (v nn))
-  (fun full ->
-    store_row #al #ms (sub full 0ul (size_row al)) (rowi hash 0ul);
-    store_row #al #ms (sub full (size_row al) (size_row al)) (rowi hash 1ul);
-    let h1 = ST.get() in
-    Lib.Sequence.eq_intro (as_seq h1 full) (Lib.Sequence.(as_seq h1 (gsub full 0ul (size_row al)) @| as_seq h1 (gsub full (size_row al) (size_row al))));
-    let final = sub full (size 0) nn in
-    copy output final)
+  let h0 = ST.get() in
+  push_frame ();
+  let double_row = 2ul *. size_row al in
+  let full = create double_row (u8 0) in
+  let first = sub full 0ul (size_row al) in
+  let second = sub full (size_row al) (size_row al) in
+  let row0 = rowi hash 0ul in
+  let row1 = rowi hash 1ul in
+  store_row first row0;
+  store_row second row1;
+  let h1 = ST.get() in
+  Lib.Sequence.eq_intro (as_seq h1 full)
+	(Lib.Sequence.(as_seq h1 (gsub full 0ul (size_row al)) @|
+		       as_seq h1 (gsub full (size_row al) (size_row al))));
+  let final = sub full (size 0) nn in
+  copy output final;
+  pop_frame()
```

Comment:
Is this change needed? 
Keeping the application of the `salloc1` combinator would be more idiomatic.

Reply:
Not needed, but the code generated from the salloc1 was a bit weird with many uu__ variables. We may be able to go back to salloc while keeping the nice C code; I was experimenting with different options and this one generated decent code, so I kept it. Indeed, using salloc for all allocations would be nice.

Reply:
I believe we could get close to the current C code with `salloc1` by introducing the same intermediate variables. The only difference would be that one cannot control the name of the allocated buffer, so `full` would become a generic `b`.
Another difference, compared to the previous code is that `salloc1` would zero the contents of the buffer in the stack when done. This still uses `memset`, so might not be honored by the compiler, but we will replace it with a more robust version eventually.

Reply:
I don't see any `uu_` variables in the previous code, only common subexpressions that could be assigned to intermediate variables.

Reply:
Let me try to do another pass with salloc and see what happens.
Regarding salloc, I'd like to make some documentation on when and how to use it.
If we think this is a good style, we may as well do it more universally around hacl*.

Reply:
fixed, and green, it seems. let's merge?

Reply:
There is some documentation in the tutorial: https://github.com/project-everest/hacl-star/blob/dev/lib/tutorial/Tutorial.fst#L260